File: examples/javascript/simple_websocket.js
/**
 * Simple WebSocket Operations using Direct API
 *
 * Demonstrates WebSocket connection and querying
 * using raw WebSocket API - no client library required
 */

const dotenv = require("dotenv");
const WebSocket = require("ws");

dotenv.config();

const BASE_URL = process.env.API_BASE_URL || "http://localhost:8080";
const WS_URL = process.env.WS_BASE_URL || "ws://localhost:8080";
const API_KEY = process.env.API_BASE_KEY || "a-test-api-key-from-ekodb";

async function getAuthToken() {
  const response = await fetch(`${BASE_URL}/api/auth/token`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ api_key: API_KEY }),
  });

  if (!response.ok) {
    throw new Error(`Auth failed: ${response.status}`);
  }

  const data = await response.json();
  return data.token;
}

async function insertTestData(token) {
  const response = await fetch(`${BASE_URL}/api/insert/websocket_test`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      name: "WebSocket Test Record",
      value: 42,
      active: true,
    }),
  });

  if (!response.ok) {
    throw new Error(`Insert failed: ${response.status}`);
  }

  return response.json();
}

async function deleteTestCollection(token) {
  const response = await fetch(`${BASE_URL}/api/collections/websocket_test`, {
    method: "DELETE",
    headers: { Authorization: `Bearer ${token}` },
  });
  if (!response.ok && response.status !== 404) {
    throw new Error(`Collection cleanup failed: ${response.status}`);
  }
}

async function main() {
  console.log("=== Simple WebSocket Operations (Direct API) ===\n");

  // Step 1: Get authentication token
  const token = await getAuthToken();
  console.log("✓ Authentication successful");
  await deleteTestCollection(token);

  // Step 2: Insert test data first
  console.log("\n=== Inserting Test Data ===");
  const insertResult = await insertTestData(token);
  console.log(`✓ Inserted test record: ${insertResult.id}`);

  // Step 3: Connect to WebSocket
  console.log("\n=== Connecting to WebSocket ===");

  await new Promise((resolve, reject) => {
    const ws = new WebSocket(`${WS_URL}/api/ws`, {
      headers: {
        Authorization: `Bearer ${token}`,
      },
    });

    ws.on("open", () => {
      console.log("✓ WebSocket connected");

      // Step 4: Query data via WebSocket
      console.log("\n=== Querying Data via WebSocket ===");
      const messageId = Date.now().toString();

      ws.send(
        JSON.stringify({
          type: "FindAll",
          messageId: messageId,
          payload: {
            collection: "websocket_test",
          },
        }),
      );
    });

    ws.on("message", (data) => {
      try {
        const response = JSON.parse(data.toString());
        console.log("Response:", JSON.stringify(response, null, 2));

        const records = response?.payload?.data;
        if (!Array.isArray(records)) {
          throw new Error("WebSocket response did not contain payload.data");
        }
        if (records.length !== 1) {
          throw new Error(
            `Expected exactly 1 WebSocket record, got ${records.length}`,
          );
        }
        console.log("✓ Retrieved 1 record via WebSocket");

        console.log("\n✓ WebSocket example completed successfully");
        ws.close();
        resolve();
      } catch (error) {
        console.error("Failed to parse response:", error);
        ws.close();
        reject(error);
      }
    });

    ws.on("error", (error) => {
      console.error("WebSocket error:", error);
      reject(error);
    });

    ws.on("close", () => {
      console.log("WebSocket closed");
    });

    // Timeout after 10 seconds
    setTimeout(() => {
      ws.close();
      reject(new Error("WebSocket timeout"));
    }, 10000);
  });

  await deleteTestCollection(token);
}

main().catch((error) => {
  console.error("Error:", error);
  process.exit(1);
});

File: examples/python/simple_websocket.py
#!/usr/bin/env python3
"""
Simple WebSocket Operations using Direct API

Demonstrates WebSocket connection and querying
using raw WebSocket API - no client library required
"""

import asyncio
import json
import os

import requests
import websockets
from dotenv import load_dotenv

load_dotenv()

BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8080")
WS_URL = os.getenv("WS_BASE_URL", "ws://localhost:8080")
API_KEY = os.getenv("API_BASE_KEY", "a-test-api-key-from-ekodb")
COLLECTION = "simple_websocket_example_py"


def get_auth_token():
    response = requests.post(
        f"{BASE_URL}/api/auth/token",
        json={"api_key": API_KEY},
        headers={"Content-Type": "application/json"},
    )
    response.raise_for_status()
    return response.json()["token"]


def insert_test_data(token):
    response = requests.post(
        f"{BASE_URL}/api/insert/{COLLECTION}",
        json={
            "name": "WebSocket Test Record",
            "value": 42,
            "active": True,
        },
        headers={
            "Authorization": f"Bearer {token}",
            "Content-Type": "application/json",
        },
    )
    response.raise_for_status()
    return response.json()


def delete_test_collection(token):
    response = requests.delete(
        f"{BASE_URL}/api/collections/{COLLECTION}",
        headers={"Authorization": f"Bearer {token}"},
    )
    if response.status_code != 404:
        response.raise_for_status()


async def run_example(token):
    # Step 2: Insert test data first
    print("\n=== Inserting Test Data ===")
    insert_result = insert_test_data(token)
    print(f"✓ Inserted test record: {insert_result['id']}")

    # Step 3: Connect to WebSocket
    print("\n=== Connecting to WebSocket ===")

    headers = {"Authorization": f"Bearer {token}"}

    async with websockets.connect(
        f"{WS_URL}/api/ws", additional_headers=headers
    ) as websocket:
        print("✓ WebSocket connected")

        # Step 4: Query data via WebSocket
        print("\n=== Querying Data via WebSocket ===")
        message_id = str(int(asyncio.get_event_loop().time() * 1000))

        await websocket.send(
            json.dumps(
                {
                    "type": "FindAll",
                    "messageId": message_id,
                    "payload": {
                        "collection": COLLECTION,
                    },
                }
            )
        )

        # Wait for response
        try:
            response_text = await asyncio.wait_for(websocket.recv(), timeout=10.0)
            response = json.loads(response_text)
            print(f"Response: {json.dumps(response, indent=2)}")

            records = response.get("payload", {}).get("data")
            if not isinstance(records, list):
                raise RuntimeError("WebSocket response did not contain payload.data")
            if len(records) != 1:
                raise RuntimeError(
                    f"Expected exactly 1 WebSocket record, got {len(records)}"
                )
            print("✓ Retrieved 1 record via WebSocket")

        except asyncio.TimeoutError:
            raise RuntimeError("WebSocket timeout")


async def main():
    print("=== Simple WebSocket Operations (Direct API) ===\n")

    # Step 1: Get authentication token
    token = get_auth_token()
    print("✓ Authentication successful")
    delete_test_collection(token)

    operation_error = None
    try:
        await run_example(token)
    except Exception as error:
        operation_error = error

    cleanup_error = None
    try:
        delete_test_collection(token)
    except Exception as error:
        cleanup_error = error

    if operation_error is not None:
        if cleanup_error is not None:
            operation_error.add_note(f"Cleanup also failed: {cleanup_error}")
        raise operation_error
    if cleanup_error is not None:
        raise cleanup_error

    print("\n✓ WebSocket example completed successfully")


if __name__ == "__main__":
    asyncio.run(main())
